File: src/offer7.Construct_Binary_Tree/construct_binary_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
        if(pre.size() == 0 || vin.size() != pre.size()) return nullptr;
        return Construct(pre, vin, 0, pre.size()-1, 0, vin.size()-1);
    }
    TreeNode* Construct(vector<int>& pre, vector<int>& vin, int spre, int epre, int svin, int evin){
        int rootval = pre[spre];
        TreeNode* root = new TreeNode(rootval);
        if(spre == epre) return root;
        int vinroot = svin;
        while(vin[vinroot] != rootval){
            vinroot++;
        }
        int leftlength = vinroot - svin;
        if(leftlength > 0){
            root->left = Construct(pre, vin, spre+1, spre+leftlength, svin, vinroot-1);
        }
        int rightlength = evin - vinroot;
        if(rightlength > 0){
            root->right = Construct(pre, vin, spre+leftlength+1, epre, vinroot+1, evin);
        }
        return root;
    }
};

File: src/offer07.Construct_Binary_Tree/construct_binary_tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode reConstructBinaryTree(int [] pre,int [] in) {
        return helper(pre, in, 0, pre.length-1, 0, in.length-1);
    }
    private TreeNode helper(int[] pre, int[] in, int preStart, int preEnd, int inStart, int inEnd){
        if(preEnd < preStart || inEnd < inStart){
            return null;
        }
        TreeNode root = new TreeNode(pre[preStart]);
        if(preStart == preEnd){
            return root;
        }
        int index = inStart;
        while(pre[preStart] != in[index]){
            index++;
        }
        root.left = helper(pre, in, preStart+1, preStart + 1 + index-1-inStart, inStart, index-1);
        root.right = helper(pre, in, preStart + 1 + index-1-inStart + 1, preEnd, index+1, inEnd);
        return root;
    }
}